### ISSUE TITLE
what  do or_var  and  if_var mean

### ISSUE THREAD

**Issue by lihuahua123** [Bracktus/Life-Reverser#1]
or_var = or_variables[r][c] # 1 if max is 1 or 2, 
                                                  # 0 if max = 9

 if_var = if_variables[r][c] # 1 if or_var implies cell
                                                # 0 otherwise

**Comment by Bracktus** [comment]
Hi,

I agree `if_var` and `or_var` are quite confusing. 

I'm very new to linear programming so this may be riddled with mistakes 😬 but I'll try and explain my thinking.

If a cell is dead and was previously alive (and we're not on the corner or an edge) then it must've had 0, 1 or 4+ neighbours. Or to put it another way:

0 >= neighbours >= 1 OR
4 >= neighbours >= 8

(8 is assuming that we're not on the edge or a corner).

We can encode 0 >= neighbours >= 1 by doing 

`prob += lpSum(neighbours) >= 1 - cell` and
`prob += lpSum(neighbours) <= 1` because

when cell = 0, then 1 - cell = 1 and
when cell = 1, then 1 - cell = 0

and we can encode 4 >= neighbours >= 8 by doing

`probs += lpSum(neigbours) >= 4` and
`probs += lpSum(neigbours) <= 8`.

But in order to encode the choice between 0,1 OR 4 we need to introduce `or_var`, so the encoding becomes 

`prob += lpSum(neighbours) >= 4 * (1 - or_var)` and
`prob += lpSum(neighbours) <= 1 + 7 * (1 - or_var)` because

| or_var | 4*(1 - or_var) | 1 + 7 * (1 - or_var) |
|--------|----------------|----------------------|
| 1      | 0              | 1                    |
| 0      | 4              | 8                    |

However, we still need to account for when the cell is dead, and was previously dead. So it must've had 0, 1, 2 or 4+ neighbours

So the formulations becomes (when we're not on a corner or an edge):

if the cell is dead and was previously alive:
0 >= neighbours >= 1 OR
4 >= neighbours >= 8

if the the cell is dead and was previously *dead*:
0 >= neighbours >= 2 OR
4 >= neighbours >= 8 

So currently, `or_var` encodes whether we have the range 0-1 or 4-8. 

So when the cell was previously dead (cell = 0) AND our range is 0-1 (or_var = 1)
then 0 >= neighbours >= 2.

For this we need a new variable that is false under these conditions.

Luckily we can use the implies boolean operator for this.

| or_var | cell | (or_var -> cell) = if_var |
|--------|------|----------------|
| T      | T    | T              |
| T      | F    | F              |
| F      | T    | T              |
| F      | F    | T              |

I found this link on stack exchange that shows us how to encode boolean logic in LP and stole it. 
https://cs.stackexchange.com/questions/12102/express-boolean-logic-operations-in-zero-one-integer-linear-programming-ilp

So now we have

if (cell is dead (cell = 0) AND range is 0-1 (or_var = 1)) or to put it another way: if (if_var = 0)

0 >= neighbours >= 2 or
4 >= neighbours >= 8

`prob += lpSum(neighbours) >= 4 * (1 - or_var)` and
`prob += lpSum(neighbours) <= 1 + (1 - if_var) + 7 * (1 - or_var)` because

| or_var | cell | if_var | 4 * (1 - or_var) | 1 + (1 - if_var) + 7 * (1 - or_var) |
|--------|------|--------|------------------|-------------------------------------|
| 1      | 1    | 1      | 0                | 1                                   |
| 1      | 0    | 0      | 0                | 2                                   |
| 0      | 1    | 1      | 4                | 8                                   |
| 0      | 0    | 1      | 4                | 8                                   |

To summarise: 
(or_var = 0) means range = [0,1]
(or_var = 1) means range = [4,8]

(if_var = 0) means (range = [0,1] and cell = dead)
(if_var = 1) means NOT(range = [0,1] and cell = dead)

I hope this helps! :D
